### PR TITLE
Updates specs regarding error buffers

### PR DESCRIPTION
**Pull request by lokokung**
- Updates documentation regarding error buffer destruction.
- Updates documentation on mappedAtCreation buffers for error cases.

***
<a href="https://pr-preview.s3.amazonaws.com/lokokung/gpuweb/pull/2314.html" title="Last updated on Dec 6, 2021, 11:16 PM UTC (cb845a7)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/2314/cd00966...lokokung:cb845a7.html" title="Last updated on Dec 6, 2021, 11:16 PM UTC (cb845a7)">Diff</a>

### CODE REVIEW

**Comment by lokokung** [comment]
Note that this is just a split of the changes in the original PR that included both spec changes (https://github.com/gpuweb/gpuweb/pull/2312).

**Review by Kangz — APPROVED** [review]

File: spec/index.bs [hunk 1]
```diff
@@ -2102,7 +2107,8 @@ namespace GPUBufferUsage {
 ## Buffer Destruction ## {#buffer-destruction}
 
 An application that no longer requires a {{GPUBuffer}} can choose to lose
-access to it before garbage collection by calling {{GPUBuffer/destroy()}}.
+access to it before garbage collection by calling {{GPUBuffer/destroy()}}. Destroying error
+buffers is also valid as it will attempt to reclaim the memory allocated for the temporary buffer.
```
[the remark this answers is not in the document]
Comment (lokokung) [reply spec/index.bs]: Done

File: spec/index.bs [hunk 2]
```diff
@@ -2076,6 +2074,13 @@ namespace GPUBufferUsage {
             1. Set |b|.{{GPUBuffer/[[usage]]}} to |descriptor|.{{GPUBufferDescriptor/usage}}.
             1. If |descriptor|.{{GPUBufferDescriptor/mappedAtCreation}} is `true`:
 
+                1. If |b| is an error buffer:
+
+                    1. Try to allocate a temporary buffer to appear "mapped". If successfully allocated,
+                         set |b|.{{GPUBuffer/[[state]]}} to [=buffer state/mapped at creation=], otherwise
+                         Set |b|.{{GPUBuffer/[[state]]}} to [=buffer state/unmapped=].
```
[the remark this answers is not in the document]
Comment (lokokung) [reply spec/index.bs]: Ack

File: spec/index.bs [hunk 3]
```diff
@@ -2076,6 +2074,13 @@ namespace GPUBufferUsage {
             1. Set |b|.{{GPUBuffer/[[usage]]}} to |descriptor|.{{GPUBufferDescriptor/usage}}.
             1. If |descriptor|.{{GPUBufferDescriptor/mappedAtCreation}} is `true`:
 
+                1. If |b| is an error buffer:
```
[the remark this answers is not in the document]
Comment (lokokung) [reply spec/index.bs]: Done

**Comment by lokokung** [comment]
Sorry it took so long to get back to this! Updated according to the comments and our offline conversation.

File: spec/index.bs [hunk 4]
```diff
@@ -2066,11 +2069,13 @@ namespace GPUBufferUsage {
                     - If |descriptor|.{{GPUBufferDescriptor/mappedAtCreation}} is `true`:
                         - |descriptor|.{{GPUBufferDescriptor/size}} is a multiple of 4.
 
-                    Issue(gpuweb/gpuweb#605): Explain that the resulting error buffer can still be mapped at creation.
-
                     Issue(gpuweb/gpuweb#605): Explain what are a {{GPUDevice}}'s `[[allowed buffer usages]]`.
                 </div>
 
+            Note: If buffer creation fails, and |descriptor|.{{GPUBufferDescriptor/mappedAtCreation}} is `false`,
```
Comment (kvark) [inline spec/index.bs:2075]: I don't quite understand who is this note targeted to.
Does it inform the user in any actionable way?
Does it inform the implementors?

Comment (kainino0x) [reply spec/index.bs:2075]: I think it informs both:
Users know they don't have to do anything special to free shadow resources for failed buffers.
Implementers know that's when they should free shadow copies/etc made for mapping.

**Review by kainino0x — APPROVED** [review]

File: spec/index.bs (hunk 4 above)
Comment (kvark) [reply spec/index.bs:2075]: > Users know they don't have to do anything special to free shadow resources for failed buffers.

I don't think anybody would assume that anything has to be done in the first place.

Comment (kainino0x) [reply spec/index.bs:2075]: Maybe not at first, but a user who is a little bit familiar with the implementation architecture (which is a good chunk of users, usually) may well be unsure. In any case, I think it still informs implementers, so shouldn't we keep it for that reason?

Comment (kvark) [reply spec/index.bs:2075]: This implementation behavior is non-observable, and I don't even see this as something our implementation would follow.

Comment (kvark) [reply spec/index.bs:2075]: How do you even `destroy()` a failed buffer? A failed buffer shouldn't support any operations on it, anyway.

Comment (kainino0x) [reply spec/index.bs:2075]: I don't think destroying a failed buffer should do anything, but I also think we should allow it so applications don't have to special-case their cleanup logic.

File: spec/index.bs [hunk 5]
```diff
@@ -2066,11 +2069,13 @@ namespace GPUBufferUsage {
...
+            it is impossible to call {{GPUBuffer/mapAsync()}}, so any resources allocated to enable mapping can
```
Comment (toji) [inline spec/index.bs:2076]:
```suggestion
            any calls to {{GPUBuffer/mapAsync()}} will reject, so any resources allocated to enable mapping can
```

**Review by kvark — APPROVED** [review]

**Review by toji — APPROVED** [review]

**Comment by kainino0x** [comment]
in the editors meeting @kvark suggested maybe we should have a single place that describes the interactions between mapping and invalidity. This is probably a good idea but I don't think it blocks this PR.

**Comment by kainino0x** [comment]
Oops forgot to commit the suggestion.

**Pull request merged**